File: opswat.py
import hashlib
import json
import multiprocessing
import os
import pprint
import requests
import sys
import time

"""
NOTE:
    Tested with PE files, office documents, and archive files like zip and 7z.
    The documention stated that data_id's for each file would be present in
    the scan result if an archive file was scanned, but in my testing I did
    not see these fields populated in the scan result JSON object, so I did
    not implement any code that would display scan results for each file in
    the archive.
"""

"""
Register for a free account at portal.opswat.com.
Copy and paste your API key here
"""
APIKEY = "<ENTER YOUR API KEY HERE>"


def metadefender_cloud_hash_scan(file_sha256):
    """
    Lookup scan results using file SHA256
    Input:  file SHA256
    Return: metadefender JSON response containing
            scan results, else None
    """
    headers = {
        "apikey" : APIKEY,
    }

    # Make a request to metadefender cloud to scan file SHA256
    response = requests.get("https://api.metadefender.com/v4/hash/" + file_sha256, headers=headers)
    if response.status_code == 200:
        response_json = json.loads(response.text)
        return response_json
    
    # Request failed, so return None
    # {"error":{"code":404003,"messages":["The hash was not found"]}}
    return None


def metadefender_cloud_file_scan(file_path):
    """
    Upload file to metadefender cloud
    Input:  file path
    Return: metadefender JSON response containing
            data_id
    """
    headers = {
        "apikey" : APIKEY,
        "filename" : os.path.split(file_path)[1],
        "content-type" : "application/octet-stream",
    }

    # Read file contents
    data = open(file_path, "rb").read()

    # Make a request to metadefender cloud to scan the file
    response = requests.post("https://api.metadefender.com/v4/file", headers=headers, data=data)
    if response.status_code == 200:
        response_json = json.loads(response.text)
        return response_json
    
    # Request failed, so return None
    return None


def metadefender_cloud_file_scan_poll(data_id, timeout=60):
    """
    Poll metadefender cloud for scan results
    Input:  data_id of uploaded file
            timeout value in seconds (default is 60)
    Return: metadefender JSON response containing
            scan results, else None
    """
    headers = {
        "apikey" : APIKEY,
    }

    cnt = 0
    while cnt < timeout / 10:
        # Make a request to metadefender cloud for scan progress
        response = requests.get("https://api.metadefender.com/v4/file/" + data_id, headers=headers)
        if response.status_code == 200:
            response_json = json.loads(response.text)
            if response_json["scan_results"]["progress_percentage"] == 100:
                # scan is done, so return the scan results
                return response_json
            # sleep for 10 seconds before making the next request
            time.sleep(10)
            cnt += 1
    
    return None


def display_scan_result(scan_result):
    """
    Display the metadefender scan results
    Input: scan results
    Output: output engine scan results to stdout
    """
    filename = scan_result["file_info"]["display_name"]
    overall_status = scan_result["scan_results"]["scan_all_result_a"]

    print(f"filename: {filename}")
    print(f"overall_status: {overall_status}")

    for scan_engine in scan_result["scan_results"]["scan_details"]:
        print(f"engine: {scan_engine}")

        threat_found = scan_result["scan_results"]["scan_details"][scan_engine]["threat_found"]
        print(f"threat_found: {threat_found}")

        scan_result_i = scan_result["scan_results"]["scan_details"][scan_engine]["scan_result_i"]
        print(f"scan_result: {scan_result_i}")

        def_time = scan_result["scan_results"]["scan_details"][scan_engine]["def_time"]
        print(f"def_time: {def_time}")


def get_file_sha256(file_path):
    """
    Compute file SHA256 of file
    Input: file path
    Return: SHA256 string
    """
    if not os.path.exists(file_path):
        return None

    sha256_hash = hashlib.sha256()
    with open(file_path, "rb") as f:
        for byte_block in iter(lambda: f.read(4096), b""):
            sha256_hash.update(byte_block)
    return sha256_hash.hexdigest()


def main(file_path):
    # Get file SHA256
    file_sha256 = get_file_sha256(file_path)
    if not file_sha256:
        print(f"ERROR: {file_path} does not exist")
        return

    print(f"INFO: SHA256 of {file_path} is {file_sha256}")

    scan_result = None
    try:
        # Scan the file SHA256
        scan_result = metadefender_cloud_hash_scan(file_sha256)
        if not scan_result:
            print(f"INFO: Metadefender does not have a record of {file_sha256}")
            print(f"INFO: Uploading {file_sha256} to Metadefender cloud")

            # File SHA256 does not exist on metadefender cloud, so
            # upload the file
            scan_result = metadefender_cloud_file_scan(file_path)
            if scan_result and scan_result["status"] == "inqueue":
                print(f"INFO: {file_sha256} added to the scan queue")
                print(f"INFO: data_id = {scan_result['data_id']}")
                print("INFO: Polling metadefender cloud for scan results")

                # The file has been added to the queue, so
                # now let's just poll metadefender for scan result
                scan_result = metadefender_cloud_file_scan_poll(scan_result["data_id"])

    except requests.exceptions.ConnectionError:
        print(f"ERROR: Check your network connection")
        scan_result = None

    if not scan_result:
        print("ERROR: Failed to scan the file")
    else:
        display_scan_result(scan_result)

if __name__ == "__main__":
    if len(sys.argv) == 1:
        print("USAGE: python opswat.py [path to file]")
    else:
        main(sys.argv[1])

File: opswat.go
package main

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"
	"time"
)

// Define our scan object
type ScanObject struct {
	apikey      string
	filePath    string
	fileSha256  string
	dataId      string
	scanDetails map[string]interface{}
}

func (scanObj *ScanObject) computeFileSha256() error {
	// Check if the file exists
	if _, err := os.Stat(scanObj.filePath); err != nil {
		return errors.New("file does not exist")
	}

	// Open the file
	file, err := os.Open(scanObj.filePath)
	if err != nil {
		return err
	}
	defer file.Close()

	// Create a SHA256 object
	hash := sha256.New()
	// Calculate the SHA256 for the file
	if _, err := io.Copy(hash, file); err != nil {
		return err
	}

	// Save the file SHA256 as a string in our scan object
	// We will use it to scan the SHA256 using metadefender hash scan API
	scanObj.fileSha256 = hex.EncodeToString(hash.Sum(nil))
	return nil
}

func (scanObj *ScanObject) scanSha256() error {
	// Create a request to do a metadefender hash scan on our SHA256
	req, err := http.NewRequest("GET", "https://api.metadefender.com/v4/hash/"+scanObj.fileSha256, nil)
	if err != nil {
		return err
	}

	// Add the apikey to the header
	req.Header.Add("apikey", scanObj.apikey)

	// Creae a client and make the request
	client := &http.Client{}
	res, err := client.Do(req)
	if err != nil {
		return err
	}

	defer res.Body.Close()

	// Read the response from the server
	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		return err
	}

	// Parse the response as a JSON data
	var mdRes map[string]interface{}
	err = json.Unmarshal([]byte(body), &mdRes)
	if err != nil {
		return err
	}

	// Check if "error" field is in the JSON data
	if _, found := mdRes["error"]; found {
		return errors.New("file is not in metadefender database")
	}

	// Save the scan_details in our scan object, so we
	// can use it to display to stdout
	scanRes := mdRes["scan_results"].(map[string]interface{})
	scanObj.scanDetails = scanRes["scan_details"].(map[string]interface{})
	return nil
}

func (scanObj *ScanObject) uploadScanFile() error {
	// Stole some code from the following URL
	// https://matt.aimonetti.net/posts/2013-07-golang-multipart-file-upload-example/

	// Open the file
	file, err := os.Open(scanObj.filePath)
	if err != nil {
		return err
	}

	defer file.Close()

	// Allocate a buffer for multipart/form-data
	body := &bytes.Buffer{}
	// Create a writer to our buffer
	writer := multipart.NewWriter(body)
	// Tell the writer to create the multipart/form-data field
	part, err := writer.CreateFormFile("file", filepath.Base(scanObj.filePath))
	if err != nil {
		return err
	}

	// Populate the multipart/form-data field
	_, err = io.Copy(part, file)
	if err != nil {
		return err
	}

	err = writer.Close()
	if err != nil {
		return err
	}

	// Create a post request to call the metadefender upload API
	// Set the body of the request with the multipart/form-data
	req, err := http.NewRequest("POST", "https://api.metadefender.com/v4/file", body)
	if err != nil {
		return err
	}

	// Add the necessary headers
	req.Header.Add("apikey", scanObj.apikey)
	req.Header.Set("content-type", writer.FormDataContentType())

	// Create a HTTP client and make the request
	client := &http.Client{}
	res, err := client.Do(req)
	if err != nil {
		return err
	}

	defer res.Body.Close()

	// Get the response from the server
	resBody, err := ioutil.ReadAll(res.Body)
	if err != nil {
		return err
	}

	// Parse the response body as JSON data
	var mdRes map[string]interface{}
	err = json.Unmarshal([]byte(resBody), &mdRes)
	if err != nil {
		return err
	}

	// Check if "error" is a field in the JSON data
	if _, found := mdRes["error"]; found {
		// "error" is found, so let's return the error
		return errors.New("file is not in metadefender database")
	}

	// Save the data_id to our scan object, so we can use it
	// for polling the scan results
	scanObj.dataId = mdRes["data_id"].(string)
	return nil
}

func (scanObj *ScanObject) pollScanResult() error {
	// Create a GET request to poll scan results using the data_id we got from
	// the server on the upload API call
	req, err := http.NewRequest("GET", "https://api.metadefender.com/v4/file/"+scanObj.dataId, nil)
	if err != nil {
		return err
	}

	// Add our apikey to the header
	req.Header.Add("apikey", scanObj.apikey)

	// Create a HTTP client and make the request
	client := &http.Client{}
	res, err := client.Do(req)
	if err != nil {
		return err
	}

	defer res.Body.Close()

	// Counter for the infinite loop below
	// If count == 6, return an error
	count := 0

	for {
		// Read the response body
		body, err := ioutil.ReadAll(res.Body)
		if err != nil {
			return err
		}

		// Parse the response body as JSON data
		var mdRes map[string]interface{}
		err = json.Unmarshal([]byte(body), &mdRes)
		if err != nil {
			return err
		}

		// Get the scan_results field from the JSON object
		scanRes := mdRes["scan_results"].(map[string]interface{})
		// Get the progress_percentage field from the scan_results field
		progPercent := scanRes["progress_percentage"].(float64)

		// Check if the progress_percentage field value is 100
		if progPercent == 100 {
			// Save the scan_details field in our scan object
			scanObj.scanDetails = scanRes["scan_details"].(map[string]interface{})
			break
		}

		// Sleep for 10 seconds before trying again
		time.Sleep(10 * time.Second)

		if count == 6 {
			return errors.New("failed to get scan result within a 60 seconds")
		}

		// Make the request
		res, err = client.Do(req)
		if err != nil {
			return err
		}
	}

	return nil
}

func (scanObj *ScanObject) displayScanResult() {
	// Enumerate the scan_details field and display
	// all the scan results from the various
	// scan engines that metadefender uses
	for key, val := range scanObj.scanDetails {
		fmt.Println("Scan Engine: " + key)

		scanEngDetails := val.(map[string]interface{})

		fmt.Printf("%s: %s\n", "threat_found", scanEngDetails["threat_found"].(string))
		fmt.Printf("%s: %.f\n", "scan_time", scanEngDetails["scan_time"].(float64))
		fmt.Printf("%s: %.f\n", "scan_result_i", scanEngDetails["scan_result_i"].(float64))
		fmt.Printf("%s: %s\n", "def_time", scanEngDetails["def_time"].(string))
	}
}

func main() {
	// Parse the command line
	apiKey := flag.String("k", "", "API key")
	filePath := flag.String("f", "", "File")
	flag.Parse()

	// Initialize our file object
	scanObj := ScanObject{filePath: *filePath, apikey: *apiKey}

	fmt.Println("INFO: file = " + *filePath)

	if err := scanObj.computeFileSha256(); err != nil {
		// Failed to compute SHA256
		log.Fatal(err)
	} else {
		fmt.Println("INFO: SHA256 = " + scanObj.fileSha256)

		if err := scanObj.scanSha256(); err != nil {
			// File does not exist in Metadefender database,
			// so let's upload the file
			if err := scanObj.uploadScanFile(); err != nil {
				// File upload failed
				log.Fatal(err)
			} else {
				// File upload succeeded, so let's poll
				// Metadefender cloud for scan result
				if err := scanObj.pollScanResult(); err != nil {
					// Polling for scan result failed
					log.Fatal(err)
				} else {
					// Display the scan result
					scanObj.displayScanResult()
				}
			}
		} else {
			// File SHA256 scan succeeded, so let's display
			// the scan result
			scanObj.displayScanResult()
		}
	}
}
